README
  =============
  Bsign README		$Id: README,v 1.10 2002/01/29 23:29:03 elf Exp $
  =============

  by Marc Singer, elf@buici.com
  14 January 2002 (last revision) for version <<version>>

  This document serves as an orientation guide to bsign, a program for
  embedding and verifying hash values and digital signatures.


  1.  Introduction

  The original motivation for bsign came from experience maintaining
  GNU/Linux systems that ran continuously for several years, long
  enough for hardware failure to be a significant impediment to
  reliable operation.  The idea was to augment binary files with
  information necessary to verify data integrity.  The addition of
  digital signatures is an obvious extension of the same idea and
  serves the goal of verifying the origin of the hashed file.

  As of the start of 2002, there are two applications that perform
  system file integrity verification.  Both of them, tripwire and
  integrit, use a database to store hashes of each file's contents and
  other information about the file.  Bsign stores the integrity check
  data in the executables themselves.  As a result, bsign the gpg
  public key can be stored on a small, read-only medium such as a
  floppy, guaranteeing the integrity of the check mechanism.

  This release of bsign is in use on production systems.

  Mail questions and comments to the author, elf@buici.com.


  1.1.  Copyright

  The bsign program is Copyright (C) 1998,1999,2000,2001,2002 by The
  Buici Company.  It is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License as
  published by the Free Software Foundation; either version 2 of the
  License, or (at your option) any later version.


  1.2.  Sources for related programs

  o  GNUPG, the GNU Privacy Guard is the only supported software for
     creating digital signatures.

       <URL:http://www.gnupg.org/>

  o Tripwire is a program that shares some of the attributes of
    bsign.  Last time I looked, it generates a potentially large
    database of hash values for the files on a host which it kept on a
    removable medium.  I developed bsign after testing tripwire and
    finding it unsuitable.

      <URL:http://www.tripwire.com>

    Version 2.3 of tripwire is available from the Debian package
    mirrors.

      apt-get install tripwire


   o The Integrit project is hosted by sourceforge.  While I haven't
     tried it, integrit shares the database feature of tripwire making
     it unsuitable for a read-only floppy setup.
 

  1.3  The most current source for Bsign

  The primary site for Bsign source is Buici.  

    <URL:ftp://ftp.buici.com/pub/bsign>


  1.4  Feedback and Bug Reports

  Bug reports may be sent to the author at <elf@buici.com>.  There is
  a plan to incorporate some form of automatic bug report generation
  within the application, but until that is available, e-mail is
  likely to produce a response.  


  1.5  Disclaimer

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details. 


  1.6  Known Issues

  o Release 1.0.6 of gpg does not permit signature verification using
    a key on a read-only medium.  For the time being, the verification
    diskette must remain writable.
  o The ELF-64 format is unsupported.  Bsign will fail to sign a ia64
    and alpha executables.
  o Cross-platform signing works as long as the endianness of the host
    and target are the same.  Bsign will fail to sign when the
    endianness differs.
  o When gnupg fails, bsign sometimes needs to guess that the reason
    is that it isn't installed.  Error messages indicate this
    possibility.  It would be better if bsign checked for the presence
    of the program explicitly, or if bsign used libgcrypt.


  2.  Usage

  There are two modes for using bsign: verifying program integrity,
  and verifying program authenticity.  In the first mode, bsign embeds
  a checksum in each executable that is easily verified by bsign.  An
  error in the checksum indicates corruption.  In second mode, bsign
  embeds a checksum and then signs it.  The setup for signatures
  requires a few more steps in order to prevent tampering.


  2.1  Signing Executables with Bsign

  The concept of signing executable programs is simple.  Create a gpg
  key pair.  Sign executables on the target host.  Physically remove
  the secret key from the machine and leave the public key available
  on a read-only medium.  Check signatures regularly.  The devil is in
  missing the details.

  2.1.1  Prepare Key Diskettes

  The paranoid operator will want to start as cleanly as possible.
  For those running Debian, it is sufficient to install the bsign
  package.  Otherwise, download the bsign source archive from
  ftp.buici.com.  Bring the host down to single user mode.

    telinit s

  Build bsign.

    tar zxf bsign_<<version>>.tar.gz
    cd bsign-<<version>>
    configure
    make

  Format and mount a floppy diskette.  Copy the bsign executable to
  it.  Create a gpg key pair using the diskette as home.  Note that
  gpg appears to require a complete UNIX filesystem for the homedir.
  So, I format the diskette for ext2.

    superformat /dev/fd0
    mke2fs /dev/fd0
    mount /dev/fd0 /floppy
    cp bsign bsign_sign /floppy
    gpg --homedir /floppy --gen-key

  Copy the public key to the hard drive and unmount the floppy
  diskette.

    mkdir bsignkey
    cp /floppy/pubring.gpg bsignkey
    umount /floppy

  Format and mount another diskette.  Copy those files to this floppy
  as well as bsign and then unmount it.

    superformat /dev/fd0
    mke2fs /dev/fd0
    mount /dev/fd0 /floppy
    cp bsignkey/* bsign bsign_verify /floppy
    umount /floppy

  At this point, you have two diskettes.  Both have a copy of bsign.
  One contains only the public key while the other contains both the
  public and private keys.  Label the diskettes appropriately and
  toggle the read-only tabs to prevent accidental erasure.  


  2.1.2  Sign Everything

  Insert the diskette with the secret key into the floppy drive.

  You may want to edit some of the parameters in the bsign_sign
  script.  It will sign everything on the filesystem while excluding
  some of the inappropriate directories.

  When the script is ready, sign everything on the hard drive.

    /floppy/bsign_sign

  Bsign will ask key for the passphrase one time when it first signs a
  file.  This whole process may take some time.  When it finishes, it
  has undoubtedly signed /sbin/init.  It may be necessary to restart
  init after the executable has changed to prevent a unclean shutdown.
  Just in case, restart init.

    telinit q

  Unmount the diskette and store it in safe place.  You may wish to
  make a copy of the diskette, though it isn't important since the key
  itself is only used to verify the identity of the user who signed
  the executables.


  2.1.3  Setup Regular Signature Verification

  In principle, you will want to leave the public-key-only diskette in
  the floppy drive and configure a cron job to call the bsign_verify
  regularly.  With a diskette in the floppy drive, you may want to
  change the system BIOS to avoid attempts to boot the key diskette.
  This setup will verify the system every night. 

  Insert the public-key diskette into the floppy drive.  Check that
  the command is correct for your system.

  Make sure the following (or an equivalent) fstab entry exists to
  mount the floppy on access.

    /dev/fd0 /floppy auto defaults 0 0

  Test the script.

    /floppy/bsign_verify

  When it completes, check the mail that it sends.  It should not
  report any failed signatures.

  Create a crontab entry to check the signatures.  Make sure this
  is added to the root crontab.

    10 1 * * * /floppy/bsign_verify

  You will get an email report every night of the unsigned or
  invalidly signed programs.


  2.2  Hashing Executables

  Protecting a host by only hashing executables is much simpler than
  attempting to sign them.  The bsign_hash script will hash everything
  in the filesystem.

  Then, add the crontab entry to check them automatically.  On a
  Debian system, it might look like this:

    10 1 * * * /usr/share/doc/bsign/scripts/bsign_check

  
  3.  Security Considerations

  While the only secure computer is one that is turned off, the need
  for running systems overwhelms us.  The next most secure systems are
  those that eliminate all but essential services and those where
  network logins are protected by secure passwords.  Still, clear-text
  passwords will be sniffed by crackers and dissatisfied employees
  will inveigle backdoor entrances.  If a system is cracked and a
  root-kit installed, standard system utilities are replaced by
  patched program that conceal the intruder.  Bsign will detect these
  files, alerting the operator that something has changed.

  Bsign, statically linked and stored on the diskette, depends on gpg
  as an external program.  When Werner Koch finishes with his
  libgcrypt library, I'll link that in, too, so that bsign requires
  neither dynamic libraries nor external programs.

  The validation script depends on cron, and the mail subsystem.
  Problems with these will only delay discovery of a compromised host.
  As long as the secret key is intact, a system may be taken off-line
  and scanned for untrusted software.  Once found, corrupted packages
  on Debian system are easily restored

    apt-get install --reinstall <package>

  Without a tool like bsign, it is usually more time efficient to
  rebuild the host from scratch than to attempt to discover and
  erradicate untrusted software.

  Once compromised, it is important to discover the means of entry.
  Replace software with an known exploits.  In addition, the system
  password file is forfeit as well as the ssh server key.


  4.  Signature Format

  Bsign creates an ELF section at the end of the file to contain the
  hash and optional digital signature.  The name of the section is
  "signature" and the type is the four byte string 0x80, 's', 'i',
  'g'.

  The signature section begins with a newline terminated
  identification string.  The first chacter is a '#' followed by a
  version number and a semicolon.  In version 1, the rest of the line
  identifies the program, bsign, and its version number.

  The next twenty bytes are the SHA1 hash of the file contents
  including a signature section initialized to null bytes.  The hash
  is written as binary data in MSB order.

  Following the hash is a two byte, MSB ordered length of the digital
  signature, if there is one.  This signature data is in the OpenPGP
  packet format as generated by GNUPG and PGP.  At the moment, only
  GNUPG is supported.

  This format is intentionally simple since it is likely to be read
  many more times by the machine than my humans.  Signature date
  stamps, file permissions, or ownership bits are not included in this
  version, but may be added by incrementing the version number and
  restructuring the signature data.

  It may appear strange that the file is hashed and then the digital
  signature is made on this hash.  While the digital signature is
  sufficient for verifying the validity of the hash, the original
  intention of bsign is to detect file corruption.  files may be
  hashed without being digitally signed.  By using a separate hash,
  the file contents may be verified without checking the signature,
  and therefore, without the need for gpg.




;;; Local Variables: ***
;;; mode: indented-text ***
;;; End: ***


debian/changelog
bsign (0.4.5) unstable; urgency=low

  * Recast the signature data pointer to prevent inappropriate sign
    extension.  

 -- Marc Singer <elf@buici.com>  Wed,  6 Aug 2003 14:35:54 -0700

bsign (0.4.4) unstable; urgency=low

  * Removed install dependency on gpg so that bsign can install in main.  
  * Detect absence of gnupg when running commands that require it.

 -- Marc Singer <elf@buici.com>  Tue, 29 Jan 2002 15:26:00 -0800

bsign (0.4.3) unstable; urgency=low

  * Tested/fixed the bsign_ scripts.
  * Enhanced summary.

 -- Marc Singer <elf@buici.com>  Sun, 27 Jan 2002 10:46:02 -0800

bsign (0.4.2) unstable; urgency=low

  * Reopened project after long hiatus.  Need identified to use it, so
    there's a need to complete the work.  Shouldn't be too difficult since
    the signing has always worked.  It's UI polish that was missing. 
  * Fixed handling of temporary files and signals so that no temporaries
    are left behind.
  * Fixed the dup_status function so that setuid bit isn't cleared.
  * Added fts_ enumeration mode and unified all of the enumeration into a
    class. 
  * Made some changes to the file processing to avoid mmap'ing files until
    we know that we will be reading the whole file.
  * No longer resigning by default.
  * Summary option.
  * Rewrote man page to reflect new options and to be more clear. 
  * Added scripts to demonstrate use on productionsystem.
  * Improved the handling of informational messages and exit codes so that
    email of output is useful.
  * Changed semantics of command line to be more obvious.  The problem was
    that the switches were interpreted in order.  If a filename appeared
    before the command, bsign wouldn't know what command to execute.  Now,
    all options are read before starting work.
  * Fixed the replace_file function so that there is no opportunity to
    lose a file.

 -- Marc Singer <elf@buici.com>  Thu, 17 Jan 2002 14:14:08 -0800

bsign (0.2.0) unstable; urgency=low

  * Edits to README. 
  * Appear to be some bug fixes, too.

 -- Marc Singer <elf@buici.com>  Fri,  5 May 2000 09:49:16 -0700

bsign (0.1.7) unstable; urgency=low

  * Disabled echo on passphrase read.  Less paranoid about elf'ness when
    reading segments.  
  
  * bss (NOBITS) section can have any durn offset it wants. 

 -- Oscar Levi <elf@buici.com>  Mon, 14 Dec 1998 01:07:44 -0800

bsign (0.1.6) unstable; urgency=low

  * First public release.  Digital signature only by GNUPG.

 -- Oscar Levi <elf@buici.com>  Sat, 12 Dec 1998 12:14:35 -0800




sha1.c
/* sha1.c - SHA1 hash function
 *	Copyright (C) 1998 Free Software Foundation, Inc.
 *
 * Please see below for more legal information!
 *
 * This file is part of GNUPG.  It is included in BSIGN in accordance
 * with the licence agreement of GNUPG.  It has been modified little,
 * only to make it compile and link with BSIGN.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */


/*  Test vectors:
 *
 *  "abc"
 *  A999 3E36 4706 816A BA3E  2571 7850 C26C 9CD0 D89D
 *
 *  "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
 *  8498 3E44 1C3B D26E BAAE  4AA1 F951 29E5 E546 70F1
 */

/* elf 1 Dec 1998: slightly modified to fit the linkage model for bsign. */
/* elf 17 Jan 2002: changed test for host endian-ness for autoconf macro. */


#include "standard.h"
//#include <stdio.h>
//#include <stdlib.h>
//#include <string.h>
//#include <assert.h>
//#include "util.h"
//#include "memory.h"
#include "sha1.h"

#define u32 unsigned32
#define DIM(v) (sizeof(v)/sizeof((v)[0]))
#define DIMof(type,member)   DIM(((type *)0)->member)

typedef struct {
    u32  h0,h1,h2,h3,h4;
    u32  nblocks;
    byte buf[64];
    int  count;
} SHA1_CONTEXT;


#if defined(__GNUC__) && defined(__i386__)
static inline u32
rol(int n, u32 x)
{
	__asm__("roll %%cl,%0"
		:"=r" (x)
		:"0" (x),"c" (n));
	return x;
}
#else
  #define rol(n,x)  ( ((x) << (n)) | ((x) >> (32-(n))) )
#endif




void
sha1_init( SHA1_CONTEXT *hd )
{
    hd->h0 = 0x67452301;
    hd->h1 = 0xefcdab89;
    hd->h2 = 0x98badcfe;
    hd->h3 = 0x10325476;
    hd->h4 = 0xc3d2e1f0;
    hd->nblocks = 0;
    hd->count = 0;
}


/****************
 * Transform the message X which consists of 16 32-bit-words
 */
static void
transform( SHA1_CONTEXT *hd, byte *data )
{
    u32 a,b,c,d,e,tm;
    u32 x[16];

    /* get values from the chaining vars */
    a = hd->h0;
    b = hd->h1;
    c = hd->h2;
    d = hd->h3;
    e = hd->h4;

#ifdef WORDS_BIGENDIAN
    memcpy( x, data, 64 );
#else
    { int i;
      byte *p2;
      for(i=0, p2=(byte*)x; i < 16; i++, p2 += 4 ) {
	p2[3] = *data++;
	p2[2] = *data++;
	p2[1] = *data++;
	p2[0] = *data++;
      }
    }
#endif


#define K1  0x5A827999L
#define K2  0x6ED9EBA1L
#define K3  0x8F1BBCDCL
#define K4  0xCA62C1D6L
#define F1(x,y,z)   ( z ^ ( x & ( y ^ z ) ) )
#define F2(x,y,z)   ( x ^ y ^ z )
#define F3(x,y,z)   ( ( x & y ) | ( z & ( x | y ) ) )
#define F4(x,y,z)   ( x ^ y ^ z )


#define M(i) ( tm =   x[i&0x0f] ^ x[(i-14)&0x0f] \
		    ^ x[(i-8)&0x0f] ^ x[(i-3)&0x0f] \
	       , (x[i&0x0f] = (tm << 1) | (tm >> 31)) )

#define R(a,b,c,d,e,f,k,m)  do { e += rol( 5, a )     \
				      + f( b, c, d )  \
				      + k	      \
				      + m;	      \
				 b = rol( 30, b );    \
			       } while(0)
    R( a, b, c, d, e, F1, K1, x[ 0] );
    R( e, a, b, c, d, F1, K1, x[ 1] );
    R( d, e, a, b, c, F1, K1, x[ 2] );
    R( c, d, e, a, b, F1, K1, x[ 3] );
    R( b, c, d, e, a, F1, K1, x[ 4] );
    R( a, b, c, d, e, F1, K1, x[ 5] );
    R( e, a, b, c, d, F1, K1, x[ 6] );
    R( d, e, a, b, c, F1, K1, x[ 7] );
    R( c, d, e, a, b, F1, K1, x[ 8] );
    R( b, c, d, e, a, F1, K1, x[ 9] );
    R( a, b, c, d, e, F1, K1, x[10] );
    R( e, a, b, c, d, F1, K1, x[11] );
    R( d, e, a, b, c, F1, K1, x[12] );
    R( c, d, e, a, b, F1, K1, x[13] );
    R( b, c, d, e, a, F1, K1, x[14] );
    R( a, b, c, d, e, F1, K1, x[15] );
    R( e, a, b, c, d, F1, K1, M(16) );
    R( d, e, a, b, c, F1, K1, M(17) );
    R( c, d, e, a, b, F1, K1, M(18) );
    R( b, c, d, e, a, F1, K1, M(19) );
    R( a, b, c, d, e, F2, K2, M(20) );
    R( e, a, b, c, d, F2, K2, M(21) );
    R( d, e, a, b, c, F2, K2, M(22) );
    R( c, d, e, a, b, F2, K2, M(23) );
    R( b, c, d, e, a, F2, K2, M(24) );
    R( a, b, c, d, e, F2, K2, M(25) );
    R( e, a, b, c, d, F2, K2, M(26) );
    R( d, e, a, b, c, F2, K2, M(27) );
    R( c, d, e, a, b, F2, K2, M(28) );
    R( b, c, d, e, a, F2, K2, M(29) );
    R( a, b, c, d, e, F2, K2, M(30) );
    R( e, a, b, c, d, F2, K2, M(31) );
    R( d, e, a, b, c, F2, K2, M(32) );
    R( c, d, e, a, b, F2, K2, M(33) );
    R( b, c, d, e, a, F2, K2, M(34) );
    R( a, b, c, d, e, F2, K2, M(35) );
    R( e, a, b, c, d, F2, K2, M(36) );
    R( d, e, a, b, c, F2, K2, M(37) );
    R( c, d, e, a, b, F2, K2, M(38) );
    R( b, c, d, e, a, F2, K2, M(39) );
    R( a, b, c, d, e, F3, K3, M(40) );
    R( e, a, b, c, d, F3, K3, M(41) );
    R( d, e, a, b, c, F3, K3, M(42) );
    R( c, d, e, a, b, F3, K3, M(43) );
    R( b, c, d, e, a, F3, K3, M(44) );
    R( a, b, c, d, e, F3, K3, M(45) );
    R( e, a, b, c, d, F3, K3, M(46) );
    R( d, e, a, b, c, F3, K3, M(47) );
    R( c, d, e, a, b, F3, K3, M(48) );
    R( b, c, d, e, a, F3, K3, M(49) );
    R( a, b, c, d, e, F3, K3, M(50) );
    R( e, a, b, c, d, F3, K3, M(51) );
    R( d, e, a, b, c, F3, K3, M(52) );
    R( c, d, e, a, b, F3, K3, M(53) );
    R( b, c, d, e, a, F3, K3, M(54) );
    R( a, b, c, d, e, F3, K3, M(55) );
    R( e, a, b, c, d, F3, K3, M(56) );
    R( d, e, a, b, c, F3, K3, M(57) );
    R( c, d, e, a, b, F3, K3, M(58) );
    R( b, c, d, e, a, F3, K3, M(59) );
    R( a, b, c, d, e, F4, K4, M(60) );
    R( e, a, b, c, d, F4, K4, M(61) );
    R( d, e, a, b, c, F4, K4, M(62) );
    R( c, d, e, a, b, F4, K4, M(63) );
    R( b, c, d, e, a, F4, K4, M(64) );
    R( a, b, c, d, e, F4, K4, M(65) );
    R( e, a, b, c, d, F4, K4, M(66) );
    R( d, e, a, b, c, F4, K4, M(67) );
    R( c, d, e, a, b, F4, K4, M(68) );
    R( b, c, d, e, a, F4, K4, M(69) );
    R( a, b, c, d, e, F4, K4, M(70) );
    R( e, a, b, c, d, F4, K4, M(71) );
    R( d, e, a, b, c, F4, K4, M(72) );
    R( c, d, e, a, b, F4, K4, M(73) );
    R( b, c, d, e, a, F4, K4, M(74) );
    R( a, b, c, d, e, F4, K4, M(75) );
    R( e, a, b, c, d, F4, K4, M(76) );
    R( d, e, a, b, c, F4, K4, M(77) );
    R( c, d, e, a, b, F4, K4, M(78) );
    R( b, c, d, e, a, F4, K4, M(79) );

    /* update chainig vars */
    hd->h0 += a;
    hd->h1 += b;
    hd->h2 += c;
    hd->h3 += d;
    hd->h4 += e;
}


/* Update the message digest with the contents
 * of INBUF with length INLEN.
 */
static void
sha1_write( SHA1_CONTEXT *hd, byte *inbuf, size_t inlen)
{
    if( hd->count == 64 ) { /* flush the buffer */
	transform( hd, hd->buf );
	hd->count = 0;
	hd->nblocks++;
    }
    if( !inbuf )
	return;
    if( hd->count ) {
	for( ; inlen && hd->count < 64; inlen-- )
	    hd->buf[hd->count++] = *inbuf++;
	sha1_write( hd, NULL, 0 );
	if( !inlen )
	    return;
    }

    while( inlen >= 64 ) {
	transform( hd, inbuf );
	hd->count = 0;
	hd->nblocks++;
	inlen -= 64;
	inbuf += 64;
    }
    for( ; inlen && hd->count < 64; inlen-- )
	hd->buf[hd->count++] = *inbuf++;
}


/* The routine final terminates the computation and
 * returns the digest.
 * The handle is prepared for a new cycle, but adding bytes to the
 * handle will the destroy the returned buffer.
 * Returns: 20 bytes representing the digest.
 */

static void
sha1_final(SHA1_CONTEXT *hd)
{
    u32 t, msb, lsb;
    byte *p;

    sha1_write(hd, NULL, 0); /* flush */;

    msb = 0;
    t = hd->nblocks;
    if( (lsb = t << 6) < t ) /* multiply by 64 to make a byte count */
	msb++;
    msb += t >> 26;
    t = lsb;
    if( (lsb = t + hd->count) < t ) /* add the count */
	msb++;
    t = lsb;
    if( (lsb = t << 3) < t ) /* multiply by 8 to make a bit count */
	msb++;
    msb += t >> 29;

    if( hd->count < 56 ) { /* enough room */
	hd->buf[hd->count++] = 0x80; /* pad */
	while( hd->count < 56 )
	    hd->buf[hd->count++] = 0;  /* pad */
    }
    else { /* need one extra block */
	hd->buf[hd->count++] = 0x80; /* pad character */
	while( hd->count < 64 )
	    hd->buf[hd->count++] = 0;
	sha1_write(hd, NULL, 0);  /* flush */;
	memset(hd->buf, 0, 56 ); /* fill next block with zeroes */
    }
    /* append the 64 bit count */
    hd->buf[56] = msb >> 24;
    hd->buf[57] = msb >> 16;
    hd->buf[58] = msb >>  8;
    hd->buf[59] = msb	   ;
    hd->buf[60] = lsb >> 24;
    hd->buf[61] = lsb >> 16;
    hd->buf[62] = lsb >>  8;
    hd->buf[63] = lsb	   ;
    transform( hd, hd->buf );

    p = hd->buf;
#ifdef WORDS_BIGENDIAN
# define X(a) do { *(u32*)p = hd->h##a ; p += 4; } while(0)
#else /* little endian */
# define X(a) do { *p++ = hd->h##a >> 24; *p++ = hd->h##a >> 16;	 \
                   *p++ = hd->h##a >> 8; *p++ = hd->h##a; } while(0)
#endif
    X(0);
    X(1);
    X(2);
    X(3);
    X(4);
  #undef X

}

static byte *
sha1_read( SHA1_CONTEXT *hd )
{
    return hd->buf;
}

/****************
 * Return some information about the algorithm.  We need algo here to
 * distinguish different flavors of the algorithm.
 * Returns: A pointer to string describing the algorithm or NULL if
 *	    the ALGO is invalid.
 */
const char *
sha1_get_info( int algo, size_t *contextsize,
	       byte **r_asnoid, int *r_asnlen, int *r_mdlen,
	       void (**r_init)( void *c ),
	       void (**r_write)( void *c, byte *buf, size_t nbytes ),
	       void (**r_final)( void *c ),
	       byte *(**r_read)( void *c )
	     )
{
    static byte asn[15] = /* Object ID is 1.3.14.3.2.26 */
		    { 0x30, 0x21, 0x30, 0x09, 0x06, 0x05, 0x2b, 0x0e, 0x03,
		      0x02, 0x1a, 0x05, 0x00, 0x04, 0x14 };
    if( algo != 2 )
	return NULL;

    *contextsize = sizeof(SHA1_CONTEXT);
    *r_asnoid = asn;
    *r_asnlen = DIM(asn);
    *r_mdlen = 20;
    *r_init  = (void (*)(void *))sha1_init;
    *r_write = (void (*)(void *, byte*, size_t))sha1_write;
    *r_final = (void (*)(void *))sha1_final;
    *r_read  = (byte *(*)(void *))sha1_read;

    return "SHA1";
}

